### DISCUSSION TITLE
Umsetzung Duplicatus

### DISCUSSION THREAD

**Discussion by zoidable** [Plushtoast/dsa5-foundryVTT#1622] [Ideas]
Ich würde gerne Duplicatus implementieren und das Vorgehen einmal vorab abstimmen. 
Um die Umsetzung möglichst generisch vorzunehmen ist die Idee, bei den `<...>Stats` unter https://github.com/Plushtoast/dsa5-foundryVTT/blob/foundry11/modules/actor/actor-dsa5.js#L680 einen neuen Wert `hitChance` einzuführen, der standardmäßig bei 100% liegt, was bedeutet, dass eine erfolgreiche Attacke oder ein Zauber trifft (wenn nicht abgewehrt/ausgewichen wird).

Die `hitChance` würde dann in die Berechnung des `successLevel` unter https://github.com/Plushtoast/dsa5-foundryVTT/blob/foundry11/modules/system/dice-dsa5.js#L209 bzw. https://github.com/Plushtoast/dsa5-foundryVTT/blob/foundry11/modules/system/dice-dsa5.js#L1009 einfließen und könnte durch Statuseffekte beeinflusst werden. So würde ein Duplicatus mit 2 Doppelgängern die `hitChance` auf 33% senken. Entsprechende `description`s müssten ergänzt werden. 

Wie siehst du das @Plushtoast ?

**Comment by zoidable** [comment]
Alternativ könnte natürlich auch eine spezifische Implementierung von Duplicatus erfolgen, d.h. an den entsprechenden Stellen in `dice-dsa5.js` nur auf einen ActiveEffect `Duplicatus <1|2|3|4>` des Targets geprüft werden. Ich bin mir nicht sicher, ob es weitere Effekte/Regeln gibt, die die Trefferchance beeinflussen und eine generische Implementierung rechtfertigen würden.

**Comment by Plushtoast** [comment]
Gute idee und trífft sich sehr gut. der zeitpunkt wäre genau richtig, da ich das jetzt eh mal anpacken wollte

die selben fragen, hab ich mir auch schon gestellt.
grundsätzlich sehe ich die gleichen optionen bzw:

1. hard code wie du es schreibst --> am einfachsten und ich kenne auch keine andere fähigkeit mit dem gleicchen effekt
2. hitchance attribute numerisch --> bisschen problematisch wie man das im finalen wurf mit anzeigt, dazu gibt es aber ein "others" field wo man vlt einfach text dazu kloppen kann  "Trefferchance x: Der Schlag geht ins Leere"
3. die dritte option ist es, es unter situationalmodifiers verfpgbar zu machen und den active effect mit mode custom und zb "Duplicatus 1d4" oder "Duplicatus 25" anzulegen. Dann muss man anschließend diesen Wert in den situationsbedingten modifikatoren anzeigen (auch optional auswählbar zb) und später auswerten, das ist die komplizierteste lösung erlaubt aber die meiste flexibilität.

wenn du dich drum kümmern magst sind mir aber alle optionen recht, weil wie du sagst mir auch kein ähnlicher sonstiger effekt bekannt ist

**Comment by zoidable** [comment]
Option 3 verstehe ich noch nicht so ganz, aber ich schau mal was ich erreichen kann. Tendiere zu 1.

**Comment by zoidable** [comment]
Ich hab ein bischen experimentiert und das funktional schon soweit umgesetzt, bin mir allerdings noch nicht sicher wo und wie im Actor der Effekt abgelegt werden soll, also was in `changes` als `keys` steht. Aktuell wäre das `system.condition.doppelganger` mit Wert 1-4. 

Duplicatus hat ja sozusagen 4 Stufen (max. 4 Doppelgänger) - wie wäre es, das als Zustand zu implementieren? Diese Regel, dass man bei 8 Zustandsstufen insgesamt handlungsunfähig wird ist durch Zustände wie Animosität inzwischen eh überholt, oder?

**Comment by Plushtoast** [comment]
Würde das zumindest nicht doppelganger nennen. Die system.condition dinge sind speziell für etwas anderes gedacht und dann musst die auch noch anderswo behandeln sonst geben sie nen malus auf jeden einzelnen wurf. Ohne active effect könnte reichen schon mit value und max flag.oder der trefferchance key

**Comment by zoidable** [comment]
Verwende jetzt einen ActiveEffect mit `system.extra.duplicatus` als Key und der Zahl der Doppelgänger als Wert, kann natürlich noch geändert werden. Ich mach nachher mal einen PR.

**Comment by zoidable** [comment]
Hier ein entsprechendes Macro:
```
const lang = game.i18n.lang == "de" ? "de" : "en"
const dict = {
    de: {
        dialog: "Wähle die Anzahl der Doppelgänger",
        description: "{amount} Doppelgänger (Trefferchance {hitChance}% / 1-{rollTarget} auf W20)"
    },
    en: {
        dialog: "Choose the number of doppelgangers",
        description: "{amount} doppelgangers, (chance to be hit {hitChance}% / 1-{rollTarget} with d20)"
    }
}[lang]

const rounds = qs == 6 ? qs * 6 : qs * 3

const applyDuplicatus = async (amount) => {
    const hitChance = Math.floor((1 / (amount + 1)) * 100)
    const condition = {
        label: source.name,
        icon: "icons/svg/aura.svg",
        changes: [
            { key: "system.extra.duplicatus", mode: 2, value: amount }
        ],
        duration: {
            rounds: rounds
        },
        flags: {
            dsa5: {
                value: null,
                editable: true,
                customizable: true,
                description: game.i18n.format(dict["description"], { amount: amount, hitChance: hitChance, rollTarget: Math.round(hitChance / 100 * 20) }),
                hideOnToken: false,
                custom: true
            },
        },
    }
    await actor.addCondition(condition)
    await sourceActor.applyMana(amount * 4, "AsP")
}

const dialog = {
    title: source.name,
    content: dict["dialog"],
    default: '1',
    buttons: {}
}

for (let i = 1; i <= Math.min(Math.max(qs - 1, 1), 4); i++) {
    dialog.buttons[i] = {
        icon: '',
        label: String(i),
        callback: async () => {
            applyDuplicatus(i)
        }
    }
}

new Dialog(dialog).render(true)
```
edit: kleine Verbesserungen

**Comment by Plushtoast** [comment]
Kannst du mich ggf mal auf Discord kontaktieren?
